Library/OcCpuLib/OcCpuLib.c

```diff
@@ -1004,109 +1004,116 @@ ScanIntelProcessor (
+    DEBUG ((DEBUG_INFO, "OCCPU: Timer address is %Lx from %a\n", (UINT64) TimerAddr, TimerSourceType));
+    DEBUG_CODE_END ();
+    Cpu->CPUFrequencyFromTSC = OcCalculateTSCFromPMTimer (Recalculate);
 
-  //
-  // Determine our core crystal clock frequency
-  //
-  Cpu->ARTFrequency = OcCalcluateARTFrequencyIntel (&Cpu->CPUFrequencyFromART, Recalculate);
+    //
+    // Determine our core crystal clock frequency
+    //
+    Cpu->ARTFrequency = OcCalcluateARTFrequencyIntel (&Cpu->CPUFrequencyFromART, Recalculate);
 
-  //
-  // Calculate CPU frequency based on ART if present, otherwise TSC
-  //
-  Cpu->CPUFrequency = Cpu->CPUFrequencyFromART > 0 ? Cpu->CPUFrequencyFromART : Cpu->CPUFrequencyFromTSC;
+    //
+    // Calculate CPU frequency based on ART if present, otherwise TSC
+    //
+    Cpu->CPUFrequency = Cpu->CPUFrequencyFromART > 0 ? Cpu->CPUFrequencyFromART : Cpu->CPUFrequencyFromTSC;
 
-  //
-  // Verify that our two CPU frequency calculations do not differ substantially.
-  //
-  if (Cpu->CPUFrequencyFromART > 0 && Cpu->CPUFrequencyFromTSC > 0
-    && ABS((INT64) Cpu->CPUFrequencyFromART - (INT64) Cpu->CPUFrequencyFromTSC) > OC_CPU_FREQUENCY_TOLERANCE) {
-    DEBUG ((
-      DEBUG_WARN,
-      "OCCPU: ART based CPU frequency differs substantially from TSC: %11LuHz != %11LuHz\n",
-      Cpu->CPUFrequencyFromART,
-      Cpu->CPUFrequencyFromTSC
-      ));
-  }
+    //
+    // Verify that our two CPU frequency calculations do not differ substantially.
+    //
+    if (Cpu->CPUFrequencyFromART > 0 && Cpu->CPUFrequencyFromTSC > 0
+      && ABS((INT64) Cpu->CPUFrequencyFromART - (INT64) Cpu->CPUFrequencyFromTSC) > OC_CPU_FREQUENCY_TOLERANCE) {
+      DEBUG ((
+        DEBUG_WARN,
+        "OCCPU: ART based CPU frequency differs substantially from TSC: %11LuHz != %11LuHz\n",
+        Cpu->CPUFrequencyFromART,
+        Cpu->CPUFrequencyFromTSC
+        ));
+    }
 
-  //
-  // There may be some quirks with virtual CPUs (VMware is fine).
-  // Formerly we checked Cpu->MinBusRatio > 0, but we have no MinBusRatio on Penryn.
-  //
-  if (Cpu->CPUFrequency > 0 && Cpu->MaxBusRatio > Cpu->MinBusRatio) {
-    Cpu->FSBFrequency = DivU64x32 (Cpu->CPUFrequency, Cpu->MaxBusRatio);
-  } else {
     //
-    // TODO: It seems to be possible that CPU frequency == 0 here...
+    // There may be some quirks with virtual CPUs (VMware is fine).
+    // Formerly we checked Cpu->MinBusRatio > 0, but we have no MinBusRatio on Penryn.
     //
-    Cpu->FSBFrequency = 100000000; // 100 Mhz
+    if (Cpu->CPUFrequency > 0 && Cpu->MaxBusRatio > Cpu->MinBusRatio) {
+      Cpu->FSBFrequency = DivU64x32 (Cpu->CPUFrequency, Cpu->MaxBusRatio);
+    } else {
+      //
+      // TODO: It seems to be possible that CPU frequency == 0 here...
+      //
+      Cpu->FSBFrequency = 100000000; // 100 Mhz
+    }
   }
-
   //
   // Calculate number of cores
+  // If we are under virtualization, then we should get topology from acpi as same as Penryn
```

Comment:
`If we are under virtualization, then we should get the topology from CPUID the same was as with Penryn.`